weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/BeanDefiningAnnotationProcessor.java

```diff
@@ -52,6 +56,15 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         addAnnotation(deploymentUnit, new AnnotationType(TransactionScoped.class));
         addAnnotation(deploymentUnit, new AnnotationType(VIEW_SCOPED_NAME, true));
         addAnnotation(deploymentUnit, new AnnotationType(FLOW_SCOPED_NAME, true));
+        // Per section 11.2.3 of the Jakarta REST 3.1 specification:
+        // In a product that supports CDI, implementations MUST support the use of CDI-style Beans as root resource
+        // classes, providers and Application subclasses. Providers and Application subclasses MUST be singletons or
+        // use application scope.
+        // Currently, these are not specified as @Stereotype annotations with a default scope. In a later spec this may
+        // happen, in which case these can be removed.
+        addAnnotation(deploymentUnit, new AnnotationType(PROVIDER, false));
+        addAnnotation(deploymentUnit, new AnnotationType(APPLICATION_PATH, false));
+        addAnnotation(deploymentUnit, new AnnotationType(PATH, false));
```

Comment:
@jamezp I think these names (or perhaps the AnnotationType objects), and any others handled this way in this DUP, that are not required by the CDI spec itself should be passed in by the relevant subsystem using a new method in WeldCapability.

Reply:
Would that happen early enough though? Weld needs to know the full set of bean defining annotations *before* it analyses any archive in order to determine if the archive is a bean archive in the first place.

Reply:
I could add a DUP which executes between `PARSE_CDI_BEAN_DEFINING_ANNOTATIONS` and `PARSE_WELD_DEPLOYMENT`. There would be room for 128 different processors between those so that does seem reasonable.

I'll update the PR locally and do some testing.

Reply:
@bstansberry @manovotn We could do something like this https://github.com/jamezp/wildfly/commit/986a777f224b40ed52d82c594e495f1aaa6fa6b1. It passed the single tests I added. I'll run the full test suite too and see how it does. It needs some cleanup, but that was a quick test. Using a DUP requires subsystems to have a dependency on `org.jboss.as.weld` which didn't seem right. This keeps the dependency on `org.jboss.as.weld.common`.

Reply:
@manovotn -- a subsystem attribute like James did is what I had in mind. The values get provided to the weld subsystem object before any DUP runs.

@jamezp I don't see the need for a new capability interface; the existing WeldCapability should be fine.

Reply:
> a subsystem attribute like James did is what I had in mind. The values get provided to the weld subsystem object before any DUP runs.

Got it, makes sense.

> the existing WeldCapability should be fine.

+1, let's reuse that

Reply:
Ah, right. It's a capability runtime API so it's not created and added very early. I'll update the PR as I think that makes the most sense.